### ISSUE TITLE
ARMv7l does not support NEON instruction set

### ISSUE THREAD

**Issue by techvoltage** [DynamoRIO/dynamorio#2448]
ARMv7l does not support neon instructions and hence DR crashes on it (SIGILL)
e.g.  `/root/dynamorio/core/arch/arm/arm.asm` has instructions like vst.64 and vstmdb which are NEON specific.
machine:
```
Ubuntu Trusty (14.04) on armv7l
```

**Comment by egrimley** [comment]
I'm not sure who or what defines the term `armv7l`, but I think it means something like little-endian ARMv7. ARMv7 has variants with and without NEON. To work on all of them, DynamoRIO would need to distinguish some of them at run time, probably by using AT_HWCAP.

What kind of hardware and kernel do you have? What is the output on your machine from `LD_SHOW_AUXV=1 /bin/true | grep HW` and `cat /proc/cpuinfo`?

**Comment by techvoltage** [comment]
`LD_SHOW_AUXV=1 /bin/true | grep HW`

```
AT_HWCAP:    half thumb fastmult vfp edsp thumbee vfpv3 tls idiva idivt
AT_HWCAP2:       0x0
```

`cat /proc/cpuinfo`
```
processor	: 0
model name	: ARMv7 Processor rev 2 (v7l)
BogoMIPS	: 50.00
Features	: half thumb fastmult vfp edsp thumbee vfpv3 tls idiva idivt vfpd32 lpae 
CPU implementer	: 0x56
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0x584
CPU revision	: 2

processor	: 1
model name	: ARMv7 Processor rev 2 (v7l)
BogoMIPS	: 50.00
Features	: half thumb fastmult vfp edsp thumbee vfpv3 tls idiva idivt vfpd32 lpae 
CPU implementer	: 0x56
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0x584
CPU revision	: 2

processor	: 2
model name	: ARMv7 Processor rev 2 (v7l)
BogoMIPS	: 50.00
Features	: half thumb fastmult vfp edsp thumbee vfpv3 tls idiva idivt vfpd32 lpae 
CPU implementer	: 0x56
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0x584
CPU revision	: 2

processor	: 3
model name	: ARMv7 Processor rev 2 (v7l)
BogoMIPS	: 50.00
Features	: half thumb fastmult vfp edsp thumbee vfpv3 tls idiva idivt vfpd32 lpae 
CPU implementer	: 0x56
CPU architecture: 7
CPU variant	: 0x2
CPU part	: 0x584
CPU revision	: 2

Hardware	: Marvell Armada 370/XP (Device Tree)
Revision	: 0000
Serial		: 0000000000000000
```

**Comment by egrimley** [comment]
Thanks. According to `git blame`, I added the `vst1.64`. Whoops.

I could easily be wrong, but I don't think `vstmdb` requires NEON ("Advanced SIMD").

So could you try with this patch and see whether it gets you any further? If you get a SIGILL, you could run it under GDB and find out which instruction caused it and where that instruction came from.

```
--- a/core/arch/arm/arm.asm
+++ b/core/arch/arm/arm.asm
@@ -468,8 +468,7 @@ GLOBAL_LABEL(dr_try_start:)
 GLOBAL_LABEL(dr_setjmp:)
         mov      REG_R2, ARG1
         stm      REG_R2!, {REG_R4-REG_R11, sp, lr}
-        vst1.64  {d8-d11},[REG_R2]!
-        vst1.64  {d12-d15},[REG_R2]!
+        vstmia   REG_R2!, {d8-d15}
         push     {r12,lr} /* save two registers for SP-alignment */
         CALLC1(GLOBAL_REF(dr_setjmp_sigmask), ARG1)
         mov      REG_R0, #0
@@ -483,8 +482,7 @@ GLOBAL_LABEL(dr_longjmp:)
         mov      REG_R2, ARG1
         mov      REG_R0, ARG2
         ldm      REG_R2!, {REG_R4-REG_R11, sp, lr}
-        vld1.64  {d8-d11},[REG_R2]!
-        vld1.64  {d12-d15},[REG_R2]!
+        vldmia   REG_R2!, {d8-d15}
         bx       lr
         END_FUNC(dr_longjmp)
 
```

**Comment by techvoltage** [comment]
After applying the patch, 62 out of 137 tests are failing.
Most of them SIGILL at this instruction:
`teq   r0, r11, lsr r0`

**Comment by egrimley** [comment]
That's a rather unlikely instruction: it's doing something like (r0 == (r11 >> r0)). And, of course, it has nothing to do with SIMD/NEON. So I wonder if you're looking at a T32 (Thumb) instruction decoded as A32 (ARM) or vice versa. Apparently you can force GDB to disassemble as A32 with `x/16wi address`, or as T32 with `x/16hi address`. Of course the `address` should not be in the middle of an instruction. Could you try some of those options to see if you can get a more plausible disassembly for the offending instruction and its context?

(I'm keen to help fix this problem but I'm on holiday this week so I might be a bit slow in responding.)

**Comment by techvoltage** [comment]
`x/16wi address` and `x/16hi address` produced the same results.
I decided to do a `git pull` after I realized that my repo was about a week old.
I proceeded the same way as earlier ie. applied the patch by egrimley.

This time I got an assertion error.
After `bin32/drrun -debug -loglevel 4 -c api/bin/libinscount.so -- ls` and reading the log file:
I can see this issue(Assertion error):
SYSLOG_ERROR: Application /bin/ls (13667).  Internal Error: DynamoRIO debug check failure: /root/dynamorio/core/unix/signal_linux_arm.c:90 vfp->magic == VFP_MAGIC
[log.0.15261.txt](https://github.com/DynamoRIO/dynamorio/files/1049631/log.0.15261.txt)

Couple of questions:
* d16-d31 registers are not supported with `vfp3` so what about instructions `vstmdb   sp!, {d16-d31}` and `vldmdb   sp!, {d16-d31}`?
* Shouldn't we also set -mfpu=vfp instead of -mfpu=neon

**Comment by egrimley** [comment]
I think "vfp3" (VFPv3) can come with or without d16-d31. In your case, /proc/cpuinfo mentions "vfpd32", so I think you have all 32 registers.

Yes, it would be better not to use "-mfpu=neon" when compiling the C/C++ source files, though it seems fairly unlikely that the compiler would try to use a NEON instruction, particularly in the DEBUG build, which uses -O0.

It's worth trying with and without DEBUG, in case there is a problem with something specific to the DEBUG build.

**Comment by egrimley** [comment]
It may be that on some systems the VFP register values are appearing 160 bytes later in the signal stack frame than expected. I don't yet understand why this happens but I have seen this patch make tests pass in some cases:

```
--- a/core/unix/signal_private.h
+++ b/core/unix/signal_private.h
@@ -163,10 +163,10 @@ typedef struct {
     stack_t           uc_stack;
     sigcontext_t      uc_mcontext;
     kernel_sigset_t   uc_sigmask;
-    int               sigset_ex[32 - (sizeof(kernel_sigset_t)/sizeof(int))];
+    int               sigset_ex[40 + 32 - (sizeof(kernel_sigset_t)/sizeof(int))];
     /* coprocessor state is here */
     union {
-        unsigned long uc_regspace[128] __attribute__((__aligned__(8)));
+        unsigned long uc_regspace[128 - 40] __attribute__((__aligned__(8)));
         kernel_vfp_sigframe_t uc_vfp;
     } coproc;
 # else
```

**Comment by egrimley** [comment]
The main problems here have been addressed by a96dc72 and 0b75c63.
